subsys/power/power.c

```diff
@@ -101,16 +101,39 @@ enum power_states _sys_suspend(s32_t ticks)
 	post_ops_done = 0;
 	sys_pm_notify_power_state_entry(pm_state);
 
-	if (deep_sleep) {
 #if CONFIG_DEVICE_POWER_MANAGEMENT
+#ifdef CONFIG_DEVICE_IDLE_PM
+	if (sys_pm_power_state_notify_devices(pm_state)) {
+		LOG_ERR("Notify devices system power state failed!");
+		sys_pm_notify_power_state_exit(pm_state);
+		pm_state = SYS_POWER_STATE_ACTIVE;
+		sys_pm_power_state_notify_devices(pm_state);
```

Comment:
There could be a potential problem here. Say I have 7 devices. If sys_pm_power_state_notify_devices() failed when calling device_pm_notify() on the 3rd device, I don't think we need to notify the last 4 devices that we are going back to SYS_POWER_STATE_ACTIVE.

Reply:
@vanti I will consider this more. could you please review other changes first? Thanks

Reply:
@vanti as below code, before resume we will check the state of devices. Sorry for the late response. 
```
void sys_pm_resume_devices(void)
{
        int i;

        for (i = 0; i < device_count; i++) {
                if (!device_retval[i]) {
                        int idx = device_ordered_list[i];

                        device_set_power_state(&pm_device_list[idx],
                                        DEVICE_PM_ACTIVE_STATE, NULL, NULL);
                }
        }
}
```